src/rtc_stream.cpp

```diff
@@ -0,0 +1,298 @@
+/*
+
+Copyright (c) 2019 Paul-Louis Ageneau
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/aux_/rtc_stream.hpp"
+#include "libtorrent/config.hpp"
+#include "libtorrent/error.hpp"
+
+#include <rtc/rtc.hpp>
+
+namespace libtorrent {
+namespace aux {
+
+namespace ip = boost::asio::ip;
+
+rtc_stream::rtc_stream(io_context& ioc, rtc_stream_init const& init)
+	: m_io_context(ioc)
+	, m_peer_connection(init.peer_connection)
+    , m_data_channel(init.data_channel)
+{
+	m_data_channel->onMessage([this](std::variant<rtc::binary, rtc::string> const& message) {
+		// Warning: this is called from another thread
+		std::visit([this, &message](auto const &data) {
+			char const *raw = reinterpret_cast<char const*>(data.data());
+			post(m_io_context, std::bind(&rtc_stream::on_message
+				, this
+				, error_code{}
+				, std::vector<char>(raw, raw + data.size())
```

Comment:
it's a bit unfortunate that every message has to be copied into a heap-allocated buffer. Assuming this callback is only ever called from a single thread, could you just copy the buffer directly into `m_read_buffer` instead?
there are even convenience functions in boost.asio to do this.
https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/buffer_copy.html

Reply:
This should be solved with the new callbacks for back-pressure.